### DISCUSSION TITLE
Unnecessary/redundant type hints

### DISCUSSION THREAD

**Discussion by AlexanderPodorov** [microsoft/pyright#7758] [Ideas]
Is it possible to add some kind of setting checking for unnecessary/redundant type hints? Close in behavior to the existing `reportUnnecessaryTypeIgnoreComment`.
Example:
```python
i: int = 1  # redundant type hint, int type is reliably inferred
```

**Comment by erictraut** [comment]
There's really no such thing as an "unnecessary or redundant type hint". The term "type hint" is a bit misleading, so I try to avoid using it. Type annotations are not really "hints"; they're declarations.

When you add a type annotation to a variable, you're declaring the upper bound of the type that the variable can be assigned. Without a type declaration, the variable is bound only by `object`, so it can be assigned any value.

```python
i = 1
i = "" # No type violation
i = [1, 2, 3] # No type violation

j: int = 1
j = "" # Type violation
j = [1, 2, 3] # Type violation
```

So, you can choose to use type declarations if you want to enforce the type of a variable. There are reasons to do this some times.

To better understand the concept of type declarations, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts).

**Comment by AlexanderPodorov** [comment]
Thanks for the information, that all makes sense!
I was curious just about `j: int = 1` case where a type checker can infer the type, so it's not needed to put the annotation.

**Reply by erictraut** [reply]
As I said above, it's not clear what "not needed" means here because `j: int = 1` means something different than `j = 1`. The former provides a type declaration and therefore restricts what values can be assigned to `j`. The latter does not. These are two different statements with different meanings. If your intent is to declare the type of `j` to be `int`, then the annotation is needed. If that is not your intent, it is not needed. Pyright can't guess your intent here, so there's no way for pyright to determine whether the annotation is needed or not.

For both `j: int = 1` and `j = 1`, pyright infers the type of the RHS expression as `Literal[1]`, and it narrows the type of `j` upon assignment.

```python
j: int = 1
reveal_type(j) # Literal[1]
```

**Reply by AlexanderPodorov** [reply]
Ahh, I see what you mean, thanks a lot! I'm closing the discussion.
